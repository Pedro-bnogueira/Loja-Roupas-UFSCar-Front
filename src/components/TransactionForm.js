import React, { useState, useEffect } from 'react';
import { 
  Dialog, DialogTitle, DialogContent, DialogActions, 
  TextField, Button, Box, FormControl, InputLabel, Select, MenuItem 
} from "@mui/material";
import { NumericFormat } from 'react-number-format';
import axios from 'axios'

export default function TransactionForm({ open, onClose, type, products, onSave, setSnackbar }) {
  const [productId, setProductId] = useState("");
  const [quantity, setQuantity] = useState("");
  const [price, setPrice] = useState("");
  const [supplierOrBuyer, setSupplierOrBuyer] = useState("");
  const [errors, setErrors] = useState({});

  useEffect(() => {
    if (!open) {
      // Limpar os campos quando o modal fecha
      setProductId("");
      setQuantity("");
      setPrice("");
      setSupplierOrBuyer("");
      setErrors({});
    }
  }, [open]);

  const validate = () => {
    const tempErrors = {};
    if (!productId) tempErrors.productId = "Selecione um produto.";
    if (!quantity || isNaN(quantity) || Number(quantity) <= 0) 
      tempErrors.quantity = "Informe uma quantidade válida.";
    if (!price || isNaN(price) || Number(price) <= 0) 
      tempErrors.price = "Informe um preço válido.";
    if (!supplierOrBuyer.trim()) {
      tempErrors.supplierOrBuyer = type === 'in'
        ? "Informe o fornecedor."
        : "Informe o comprador.";
    }
    setErrors(tempErrors);
    return Object.keys(tempErrors).length === 0;
  };

  const handleSave = async () => {
    if (!validate()) return;
  
    const transactionData = {
      type,                  // 'in' ou 'out'
      productId: Number(productId),
      quantity: Number(quantity),
      transactionPrice: Number(price), // Alinhado com o backend
      supplierOrBuyer: supplierOrBuyer.trim(),
    };
  
    try {
      const response = await axios.post(
        `${process.env.REACT_APP_API_URL}/api/movement/register`,
        transactionData,
        { withCredentials: true } 
      );
  
      if (response.status === 201 && response.data.transactionHistory) {
        // Atualiza o estado local com a transação retornada pelo backend
        onSave(response.data.transactionHistory);
        
        // Mostrar um Snackbar de sucesso
        setSnackbar({
          open: true,
          message: 'Transação registrada com sucesso!',
          severity: 'success',
        });
  
      } else {
        // Tratar erros retornados pelo backend
        setSnackbar({
          open: true,
          message: 'Erro ao registrar a transação.',
          severity: 'error',
        });
        console.error("Erro ao registrar transação: ", response.data);
      }
    } catch (error) {
      // Tratar erros de requisição
      setSnackbar({
        open: true,
        message: 'Erro ao registrar a transação.',
        severity: 'error',
      });
      console.error("Erro ao registrar transação:", error);
    }
  };

  return (
    <Dialog open={open} onClose={onClose} fullWidth maxWidth="sm">
      <DialogTitle>
        {type === "in" ? "Registrar Compra" : "Registrar Venda"}
      </DialogTitle>
      <DialogContent>
        <Box
          component="form"
          sx={{
            display: "flex",
            flexDirection: "column",
            gap: 2,
            marginTop: 1,
          }}
          noValidate
          autoComplete="off"
        >
          {/* Produto */}
          <FormControl fullWidth error={!!errors.productId}>
            <InputLabel id="produto-label">Produto</InputLabel>
            <Select
              labelId="produto-label"
              value={productId}
              label="Produto"
              onChange={(e) => setProductId(e.target.value)}
              inputProps={{ 'aria-label': 'Produto' }}
            >
              {products.map((prod) => (
                <MenuItem key={prod.id} value={prod.id}>
                  {prod.name} (ID: {prod.id})
                </MenuItem>
              ))}
            </Select>
            {errors.productId && (
              <span style={{ color: 'red', fontSize: 12 }}>{errors.productId}</span>
            )}
          </FormControl>

          {/* Quantidade */}
          <TextField
            label="Quantidade"
            type="number"
            value={quantity}
            onChange={(e) => setQuantity(e.target.value)}
            fullWidth
            error={!!errors.quantity}
            helperText={errors.quantity}
            inputProps={{ min: 1, 'aria-label': 'Quantidade' }}
          />

          {/* Preço */}
          <NumericFormat
            customInput={TextField}
            label="Preço Total (R$)"
            value={price}
            onValueChange={(values) => {
              const { floatValue } = values;
              setPrice(floatValue);
            }}
            thousandSeparator="."
            decimalSeparator=","
            prefix="R$ "
            decimalScale={2}
            fixedDecimalScale={true}
            fullWidth
            error={!!errors.price}
            helperText={errors.price}
            required
            inputProps={{ 'aria-label': 'Preço Total (R$)' }}
          />

          {/* Fornecedor/Comprador */}
          <TextField
            label={type === 'in' ? "Fornecedor" : "Comprador"}
            value={supplierOrBuyer}
            onChange={(e) => setSupplierOrBuyer(e.target.value)}
            fullWidth
            error={!!errors.supplierOrBuyer}
            helperText={errors.supplierOrBuyer}
            required
            inputProps={{ 'aria-label': type === 'in' ? "Fornecedor" : "Comprador" }}
          />
        </Box>
      </DialogContent>
      <DialogActions>
        <Button onClick={onClose} color="primary">
          Cancelar
        </Button>
        <Button onClick={handleSave} variant="contained" color="primary">
          Salvar
        </Button>
      </DialogActions>
    </Dialog>
  );
}