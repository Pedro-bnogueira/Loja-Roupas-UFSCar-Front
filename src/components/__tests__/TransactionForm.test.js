import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import axios from 'axios';
import TransactionForm from '../TransactionForm';

// Mock do axios para controlar as respostas da chamada de API
jest.mock('axios');

describe('TransactionForm', () => {
  const mockOnClose = jest.fn();
  const mockOnSave = jest.fn();
  const mockSetSnackbar = jest.fn();

  const mockProducts = [
    { id: 1, name: 'Produto A' },
    { id: 2, name: 'Produto B' },
  ];

  const defaultProps = {
    open: true,
    onClose: mockOnClose,
    type: 'in', // 'in' para registrar compra, 'out' para registrar venda
    products: mockProducts,
    onSave: mockOnSave,
    setSnackbar: mockSetSnackbar,
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('deve renderizar corretamente o título de acordo com o tipo (compra ou venda)', () => {
    // Teste para tipo "in"
    const { rerender } = render(<TransactionForm {...defaultProps} type="in" />);
    expect(screen.getByText('Registrar Compra')).toBeInTheDocument();

    // Teste para tipo "out"
    rerender(<TransactionForm {...defaultProps} type="out" />);
    expect(screen.getByText('Registrar Venda')).toBeInTheDocument();
  });

  it('deve exibir os campos iniciais corretamente', () => {
    render(<TransactionForm {...defaultProps} />);

    // Verifica se os campos estão na tela
    expect(screen.getByLabelText('Produto')).toBeInTheDocument();
    expect(screen.getByLabelText('Quantidade')).toBeInTheDocument();
    expect(screen.getByLabelText('Preço Total (R$)')).toBeInTheDocument();
    expect(screen.getByLabelText('Fornecedor')).toBeInTheDocument(); // pois o tipo é 'in'
  });

  it('deve chamar onClose ao clicar no botão "Cancelar"', async () => {
    render(<TransactionForm {...defaultProps} />);

    const cancelButton = screen.getByRole('button', { name: /cancelar/i });
    await userEvent.click(cancelButton);

    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });

  it('não deve enviar o formulário se os campos obrigatórios não forem preenchidos', async () => {
    render(<TransactionForm {...defaultProps} />);

    // Clica em salvar sem preencher nada
    const saveButton = screen.getByRole('button', { name: /salvar/i });
    await userEvent.click(saveButton);

    // Verifica as mensagens de erro
    expect(await screen.findByText('Selecione um produto.')).toBeInTheDocument();
    expect(screen.getByText('Informe uma quantidade válida.')).toBeInTheDocument();
    expect(screen.getByText('Informe um preço válido.')).toBeInTheDocument();
    expect(screen.getByText('Informe o fornecedor.')).toBeInTheDocument();

    // Garante que não chamou a API
    expect(axios.post).not.toHaveBeenCalled();
    // E também não chamou o onSave
    expect(mockOnSave).not.toHaveBeenCalled();
  });

  it('deve exibir erros de validação se os campos forem inválidos (quantidade e preço <= 0)', async () => {
    render(<TransactionForm {...defaultProps} />);

    // Selecionar um produto
    const produtoSelect = screen.getByLabelText('Produto');
    await userEvent.click(produtoSelect);

    // Seleciona "Produto A" (ID: 1)
    const produtoAOption = screen.getByText(/produto a \(id: 1\)/i);
    await userEvent.click(produtoAOption);

    // Digitar quantidade inválida (0)
    const quantidadeInput = screen.getByLabelText('Quantidade');
    await userEvent.type(quantidadeInput, '0');

    // Digitar preço inválido (0)
    const precoInput = screen.getByLabelText('Preço Total (R$)');
    // Aqui podemos usar fireEvent.change por ser NumericFormat
    fireEvent.change(precoInput, { target: { value: '0' } });

    // Digitar fornecedor (qualquer valor para não falhar nesse campo)
    const fornecedorInput = screen.getByLabelText('Fornecedor');
    await userEvent.type(fornecedorInput, 'Fornecedor Teste');

    // Clicar em Salvar
    const saveButton = screen.getByRole('button', { name: /salvar/i });
    await userEvent.click(saveButton);

    // Verifica mensagens de erro
    expect(await screen.findByText('Informe uma quantidade válida.')).toBeInTheDocument();
    expect(screen.getByText('Informe um preço válido.')).toBeInTheDocument();

    // API não é chamada
    expect(axios.post).not.toHaveBeenCalled();
  });

  it('deve fazer a chamada de API com dados corretos e chamar onSave quando o back-end retornar sucesso', async () => {
    // Mock da resposta de sucesso
    axios.post.mockResolvedValue({
      status: 201,
      data: {
        transactionHistory: {
          id: 999,
          productId: 1,
          quantity: 10,
          transactionPrice: 100,
          supplierOrBuyer: 'Fornecedor Teste',
        },
      },
    });

    render(<TransactionForm {...defaultProps} />);

    // Preencher produto
    const produtoSelect = screen.getByLabelText('Produto');
    await userEvent.click(produtoSelect);
    await userEvent.click(screen.getByText(/produto a \(id: 1\)/i));

    // Preencher quantidade
    const quantidadeInput = screen.getByLabelText('Quantidade');
    await userEvent.type(quantidadeInput, '10');

    // Preencher preço (usaremos fireEvent para simular mudança no campo NumericFormat)
    const precoInput = screen.getByLabelText('Preço Total (R$)');
    fireEvent.change(precoInput, { target: { value: '100' } });

    // Preencher fornecedor
    const fornecedorInput = screen.getByLabelText('Fornecedor');
    await userEvent.type(fornecedorInput, 'Fornecedor Teste');

    // Clicar em Salvar
    const saveButton = screen.getByRole('button', { name: /salvar/i });
    await userEvent.click(saveButton);

    // Verificar se chamou axios.post com os dados corretos
    expect(axios.post).toHaveBeenCalledWith(
      `${process.env.REACT_APP_API_URL}/api/movement/register`,
      {
        type: 'in',
        productId: 1,
        quantity: 10,
        transactionPrice: 100,
        supplierOrBuyer: 'Fornecedor Teste',
      },
      { withCredentials: true }
    );

    expect(mockOnSave).toHaveBeenCalledWith({
      id: 999,
      productId: 1,
      quantity: 10,
      transactionPrice: 100,
      supplierOrBuyer: 'Fornecedor Teste',
    });

    // Verifica se setSnackbar foi chamado corretamente
    expect(mockSetSnackbar).toHaveBeenCalledWith({
      open: true,
      message: 'Transação registrada com sucesso!',
      severity: 'success',
    });
  });

  it('deve exibir mensagem de erro e não chamar onSave ao receber erro do backend', async () => {
    // Mock de erro do backend (status diferente de 201)
    axios.post.mockResolvedValue({
      status: 400,
      data: { error: 'Erro ao registrar transação.' },
    });

    render(<TransactionForm {...defaultProps} />);

    // Preencher todos os campos corretamente
    const produtoSelect = screen.getByLabelText('Produto');
    await userEvent.click(produtoSelect);
    await userEvent.click(screen.getByText(/produto a \(id: 1\)/i));

    const quantidadeInput = screen.getByLabelText('Quantidade');
    await userEvent.type(quantidadeInput, '10');

    const precoInput = screen.getByLabelText('Preço Total (R$)');
    fireEvent.change(precoInput, { target: { value: '100' } });

    const fornecedorInput = screen.getByLabelText('Fornecedor');
    await userEvent.type(fornecedorInput, 'Fornecedor Teste');

    const saveButton = screen.getByRole('button', { name: /salvar/i });
    await userEvent.click(saveButton);

    expect(mockSetSnackbar).toHaveBeenCalledWith({
      open: true,
      message: 'Erro ao registrar a transação.',
      severity: 'error',
    });

    // onSave não deve ser chamado
    expect(mockOnSave).not.toHaveBeenCalled();
  });

  it('deve exibir mensagem de erro caso a chamada de API retorne rejeição (ex.: erro de rede)', async () => {
    // Mock de erro de rede
    axios.post.mockRejectedValue(new Error('Network Error'));

    render(<TransactionForm {...defaultProps} />);

    // Preencher todos os campos corretamente
    const produtoSelect = screen.getByLabelText('Produto');
    await userEvent.click(produtoSelect);
    await userEvent.click(screen.getByText(/produto a \(id: 1\)/i));

    const quantidadeInput = screen.getByLabelText('Quantidade');
    await userEvent.type(quantidadeInput, '10');

    const precoInput = screen.getByLabelText('Preço Total (R$)');
    fireEvent.change(precoInput, { target: { value: '100' } });

    const fornecedorInput = screen.getByLabelText('Fornecedor');
    await userEvent.type(fornecedorInput, 'Fornecedor Teste');

    const saveButton = screen.getByRole('button', { name: /salvar/i });
    await userEvent.click(saveButton);

    expect(mockSetSnackbar).toHaveBeenCalledWith({
      open: true,
      message: 'Erro ao registrar a transação.',
      severity: 'error',
    });

    // onSave não deve ser chamado
    expect(mockOnSave).not.toHaveBeenCalled();
  });
});