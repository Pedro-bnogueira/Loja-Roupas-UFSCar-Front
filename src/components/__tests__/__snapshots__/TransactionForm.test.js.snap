// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionForm Component matches snapshot when closed 1`] = `<DocumentFragment />`;

exports[`TransactionForm Component matches snapshot when open 1`] = `<DocumentFragment />`;
